src/index.js

```diff
@@ -89,148 +138,85 @@ function AudioPlayer (audioEngine) {
-                loader.load(sounds[storedIndex].fileUrl).then(function (audioBuffer) {
-                    storedContext.soundPlayers[storedIndex].setBuffer(new Tone.Buffer(audioBuffer));
-                });
-            }());
-        } else {
-            this.soundPlayers[index].setBuffer(new Tone.Buffer(sounds[index].fileUrl));
-        }
+    // if this sprite or clone is already playing this sound, stop it first
+    if (this.activeSoundPlayers[md5]) {
+        this.activeSoundPlayers[md5].stop();
     }
 
-};
+    // create a new soundplayer to play the sound
+    var player = new SoundPlayer();
+    player.setBuffer(this.audioEngine.audioBuffers[md5]);
+    player.connect(this.effectsNode);
+    this.pitchEffect.updatePlayer(player);
+    player.start();
 
-AudioPlayer.prototype.playSound = function (index) {
-    if (!this.soundPlayers[index]) return;
+    // add it to the list of active sound players
+    this.activeSoundPlayers[md5] = player;
 
-    this.soundPlayers[index].start();
-
-    var storedContext = this;
-    return new Promise(function (resolve) {
-        storedContext.soundPlayers[index].onEnded(resolve);
-    });
-};
+    // remove sounds that are not playing from the active sound players array
+    for (var id in this.activeSoundPlayers) {
+        if (!this.activeSoundPlayers[id].isPlaying) {
+            delete this.activeSoundPlayers[id];
+        }
+    }
 
-AudioPlayer.prototype.playNoteForBeats = function (note, beats) {
-    var sec = this.beatsToSec(beats);
-    this.audioEngine.instrumentPlayer.playNoteForSecWithInst(note, sec, this.currentInstrument);
-    return this.waitForBeats(beats);
+    return player.finished();
 };
 
 AudioPlayer.prototype.playDrumForBeats = function (drum, beats) {
     this.audioEngine.drumPlayer.play(drum, this.effectsNode);
-    return this.waitForBeats(beats);
-};
-
-AudioPlayer.prototype.waitForBeats = function (beats) {
-    var storedContext = this;
-    return new Promise(function (resolve) {
-        setTimeout(function () {
-            resolve();
-        }, storedContext.beatsToSec(beats) * 1000);
-    });
-};
-
-AudioPlayer.prototype.beatsToSec = function (beats) {
-    return (60 / this.audioEngine.currentTempo) * beats;
+    return this.audioEngine.waitForBeats(beats);
 };
 
 AudioPlayer.prototype.stopAllSounds = function () {
-    // stop all sound players
-    for (var i=0; i<this.soundPlayers.length; i++) {
-        this.soundPlayers[i].stop();
+    // stop all active sound players
+    for (var md5 in this.activeSoundPlayers) {
+        this.activeSoundPlayers[md5].stop();
     }
 
     // stop all instruments
     this.audioEngine.instrumentPlayer.stopAll();
 
     // stop drum notes
     this.audioEngine.drumPlayer.stopAll();
-
 };
 
 AudioPlayer.prototype.setEffect = function (effect, value) {
     switch (effect) {
-    case 'PITCH':
-        this.pitchEffect.set(value, this.soundPlayers);
+    case 'pitch':
```

Comment:
Unless you plan on the effects list being dynamic, I would make an "enum" for them. `ShaderManager` does this for draw modes in the renderer [here](https://github.com/LLK/scratch-render/blob/b70056019827f72f5d6451163a25dc8fb9e2f1f4/src/ShaderManager.js#L134), for example.